src/hotspot/os/posix/perfMemory_posix.cpp

```diff
@@ -752,22 +759,39 @@ static void cleanup_sharedmem_resources(const char* dirname) {
       continue;
     }
 
-    // we now have a file name that converts to a valid integer
-    // that could represent a process id . if this process id
-    // matches the current process id or the process is not running,
-    // then remove the stale file resources.
+    // We now have a file name that converts to a valid integer
+    // that could represent a process id.
     //
-    // process liveness is detected by sending signal number 0 to
-    // the process id (see kill(2)). if kill determines that the
-    // process does not exist, then the file resources are removed.
-    // if kill determines that that we don't have permission to
-    // signal the process, then the file resources are assumed to
-    // be stale and are removed because the resources for such a
+    // On Linux, we first try to flock() on the file to check
+    // for JVM processes in other containers that share the same
+    // /tmp directory as the current process. See comments in
+    // create_sharedmem_file() and is_locked_by_another_process().
+    // If it's already locked by another process, then obviously it's
+    // not stale
+    //
+    // Otherwise, if this process id matches the current process id or
+    // the process is not running, we treat it as a stale file and remove it.
+    //
+    // Process liveness is detected by sending signal number 0 to
+    // the process id (see kill(2)). If kill determines that the
+    // process does not exist, then the file is removed.
+    // If kill determines that that we don't have permission to
+    // signal the process, then the file is assumed to
+    // be stale and is removed because the files for such a
     // process should be in a different user specific directory.
```

Comment:
We will get a permission error from the `kill(pid, 0)` call only after we have successfully grabbed the flock on the file. Note that if the file was created by a live JVM process that has the flock fix (i.e., this PR), regardless of which user owns the process, we will never come to here.

That the value of the `pid` variable is misleading. It is the NSPID of another JVM that created the file. If the current JVM process runs in a different PID namespace, it cannot reliably determine whether the file is stale or not.

In general, I don't think we can trust `pid` at all when containers are involved. But that's OK -- if you want to use Java in containers that share the /tmp directory, you must upgrade to a JVM that has the flock fix. Otherwise the behavior is undefined.

Otherwise, if you are:

- Not using containers. or
- Using containers that don't share /tmp

The logic for handling the `kill(pid, 0)` error is not changed by this PR, so we are bug-for-bug compatible with older JVMs. If you think the behavior should be changed, may that should be done in a separate PR?

Or, if you run into problems like "my hsperf files are randomly deleted", a simple fix is to upgrade the JVM to one that has the flock fix :-)

Reply:
Okay, you convinced me.